docs(repo): update spectral in javascript examples

heitortsergent opened the pull request:
Fixes #2178, #2148.

**Checklist**

- [ ] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

I tried to make this document a little bit more beginner-friendly as I tried to run some of these examples myself while learning Spectral:

- Replaced all `require` statements to `import`
- Moved the mentions of CommonJS to its own section
- Added a section explaining how to load multiple rulesets using the `bundleAndLoadRuleset` method
- Added an example showing how to load an external specification file
- Added link to JS ruleset section in Custom Rulesets page

Not sure if I tried to do too many changes here, would really appreciate some feedback!

Commit:
docs(repo): modified docs/guides/3-javascript.md

Commit:
docs(repo): grammar fixes

Commit:
docs(repo): fix npm capitalization

Commit:
docs(repo): add section about CommonJS

Commit:
docs(repo): convert require statements to import

Commit:
docs(repo): add section for how to load multiple rulesets

Commit:
docs(repo): fix link reference to extends

Commit:
docs(repo): replace ES6 with ESM

P0lip commented:
@heitortsergent you could get https://app.circleci.com/pipelines/github/stoplightio/spectral/8590/workflows/3b058b3a-1fbf-4403-b316-6227242bf319/jobs/30750 step to pass?
How did you commit these changes? It's best to do that directly from the cloned project and avoid using tools that do not support git hooks.

P0lip commented on docs/guides/3-javascript.md:
````diff
@@ -108,41 +165,45 @@ const fs = {
-const { Spectral } = require("@stoplight/spectral-cli");
-const { Resolver } = require("@stoplight/json-ref-resolver");
+```js title="example-5.mjs" lineNumbers
+import { join } from "path";
````
this probably won't work, because we use `require` to load the module.
Suggest to leave CJS here for now until we support ESM for custom resolvers

P0lip commented on docs/guides/3-javascript.md:
```diff
@@ -45,55 +70,87 @@ spectral.setRuleset({
+import * as path from "node:path";
+import { join } from 'path';
+import { bundleAndLoadRuleset } from "@stoplight/spectral-ruleset-bundler/with-loader";
+import Parsers from "@stoplight/spectral-parsers"; // make sure to install the package if you intend to use default parsers!
```
```suggestion
import * as Parsers from "@stoplight/spectral-parsers"; // make sure to install the package if you intend to use default parsers!
```

P0lip commented on docs/guides/3-javascript.md:
```diff
@@ -1,28 +1,53 @@
-## Linting a YAML String
+### CommonJS and ESM
+
+The examples on this page are written in ESM. If you're using CommonJS, you have to import an additional module:
```
This is only for custom functions. I don't think we cover custom functions on this page, thus I think we can remove it

heitortsergent committed:
docs(repo): remove commonjs section, change resolver example to cjs

heitortsergent commented:
@P0lip thanks for the comments!

I used Stoplight to make those changes. The CircleCI lint error there was related to the PR title, so I updated that. Now I'm getting a different error about the commit message...

```
⧗   input: Modified docs/guides/3-javascript.md
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   scope may not be empty [scope-empty]
```

Not sure how to fix it... Should I rebase the commits and change the message so they follow the format `docs(repo): {commit_message}`?

heitortsergent commented:
I'm sorry, I just read the contributing guidelines 🤦 . Still wanna be sure this is the best way to fix the lint issues, should I rebase and fix the commit messages, and force push here?

P0lip commented:
>  Should I rebase the commits and change the message so they follow the format docs(repo): {commit_message}?

I can do that for you

heitortsergent committed:
docs(repo): prettier fixes

heitortsergent commented:
@P0lip thanks for doing the rebase! I was having some issues with `prettier` but I think I got it all sorted out now, all checks are ✅ . :)

philsturgeon commented on docs/guides/3-javascript.md:
```diff
@@ -1,28 +1,39 @@
+
+Similar to using Spectral in the CLI, there are two things you'll need to run Spectral in JS:
+
+- A string or a file representing an API specification
```
Hey hey, something we've tried to do in the past is be consistent on naming things, as OpenAPI has a million different words to mean the same thing: https://github.com/openapi-contrib/glossary 

So i try to only talk about an "API Specification" when talking about the OpenAPI Specification. When its _your_ OpenAPI we call that an API Description. There are description documents and description formats (OpenAPI, which has a specification!). 

```suggestion
- A string or a file containing your structured data (OpenAPI, AsyncAPI, Kubernetes, etc).
```

heitortsergent replied:
This is really helpful, and makes total sense! I'll try to use those terms and check that glossary when I'm making changes in the future. :)

philsturgeon commented on docs/guides/3-javascript.md:
```diff
@@ -45,55 +56,87 @@ spectral.setRuleset({
-const spectral = new Spectral();
-spectral.setRuleset(ruleset);
-// lint
+- An OpenAPI specification file in the same directory as your script named `petstore.yaml`. You can use the one found [here](https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.0/petstore.yaml).
```
```suggestion
- An OpenAPI description document in the same directory as your script named `petstore.yaml`. You can use the one found [here](https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.0/petstore.yaml).
```

Can we use literally any other example than the petstore? It's a really bad spec and its very associated with old Swagger tooling. 

There's a bunch in here you could use: https://github.com/stoplightio/public-apis

heitortsergent replied:
I thought people were still using petstore 😅 . I updated it with the Plaid API description from the repository you linked to. :)

philsturgeon replied:
Some people are but we don't have to perpetrate that. 😇

heitortsergent committed:
docs(repo): update wording when referring to OpenAPI description and spec

Co-authored-by: Phil Sturgeon <67381+philsturgeon@users.noreply.github.com>

heitortsergent committed:
docs(repo): replace petstore reference with plaid

heitortsergent commented:
@P0lip are there any other changes you would like me to make before we can get this merged? :shipit:

P0lip commented:
Looks good!
